### PR TITLE
[Feature] Support Spark catalog

### PR DESCRIPTION
**Pull request by Jay-ju**
## What type of PR is this：
- [ ] BugFix
- [X] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Previously, before read or write sr table in spark must create a  external table by passing 'starrocks.table.identifier'. this pr will solve this case. we can config some iterms as
```
spark.sql.catalog.starrocks=com.starrocks.connector.spark.catalog.StarRocksCatalog
spark.sql.catalog.starrocks.fe.http.url=10.37.42.50:8030
spark.sql.catalog.starrocks.fe.jdbc.url=jdbc:mysql://10.37.42.50:9030
spark.sql.catalog.starrocks.password=
spark.sql.catalog.starrocks.user=root
```
then. we can directly use table query sqll like that 
```
show databases;
use starrocks.db;
show tables
select * from tabl1;
insert into tabl2 select * from tab1;
```

image

## Which issues of this PR fixes ：

Fixes #

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
- [ ] I have added documentation for my new feature or new function

### CODE REVIEW

File: src/test/java/com/starrocks/connector/spark/sql/CatalogTest.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
+// Copyright 2021-present StarRocks, Inc. All rights reserved.
```
Comment (banmoy) [inline src/test/java/com/starrocks/connector/spark/sql/CatalogTest.java:1]: Please improve the test at least
1. verify the correctness
2. could run it using maven `mvn clean package -Dhttp_urls="your_fe_http" -Djdbc_urls="your_fe_jdbc"`

You can refer to `ReadWriteITTest` for an example

File: src/main/java/com/starrocks/connector/spark/rest/PartitionDefinition.java [hunk 2]
```diff
@@ -32,7 +33,7 @@
 /**
  * Starrocks RDD partition info.
  */
-public class PartitionDefinition implements Serializable, Comparable<PartitionDefinition> {
+public class PartitionDefinition implements Serializable, Comparable<PartitionDefinition>, InputPartition {
```
Comment (banmoy) [inline src/main/java/com/starrocks/connector/spark/rest/PartitionDefinition.java:36]: when will `InputPartition` be used?

File: src/main/java/com/starrocks/connector/spark/sql/connect/StarRocksConnector.java [hunk 3]
```diff
@@ -71,21 +77,47 @@ public static StarRocksSchema getSchema(StarRocksConfig config) {
...
+    public static Map<String, String> loadDatabase(StarRocksConfig config, List<String> dbNames) {
+        List<String> parameters = Arrays.asList(String.join(",", dbNames));
+        List<Map<String, String>> dbs = extractColumnValuesBySql(config, LOAD_DB_QUERY, parameters);
 
-    private static final String MYSQL_SITE_URL = "https://dev.mysql.com/downloads/connector/j/";
-    private static final String MAVEN_CENTRAL_URL = "https://repo1.maven.org/maven2/mysql/mysql-connector-java/";
+        for (Map<String, String> db : dbs) {
+            String dbName = Optional.ofNullable(db.get("SCHEMA_NAME"))
+                    .orElseThrow(() -> new StarrocksException("get Database SCHEMA_NAME error"));
+            return new DatabaseSpec(dbName).toJavaMap();
```
Comment (banmoy) [inline src/main/java/com/starrocks/connector/spark/sql/connect/StarRocksConnector.java:100]: It seems this method only return on db spec, why the input `dbNames` is a list?

**Comment by banmoy** [comment]
1. please make sure all tests pass by running
`mvn clean package -Dhttp_urls="your_fe_http" -Djdbc_urls="your_fe_jdbc"`. Currently ReadWriteITTest will fail
2. which spark version this feature could support? Now there is compilation failure against spark 3.2, 3.4, 3.5

File: src/main/java/com/starrocks/connector/spark/sql/connect/StarRocksConnector.java [hunk 4]
```diff
@@ -71,21 +77,47 @@ public static StarRocksSchema getSchema(StarRocksConfig config) {
...
+        }
 
-    // Driver name for mysql connector 5.1 which is deprecated in 8.0
-    private static final String MYSQL_51_DRIVER_NAME = "com.mysql.jdbc.Driver";
+        return dbNames;
+    }
 
-    // Driver name for mysql connector 8.0
-    private static final String MYSQL_80_DRIVER_NAME = "com.mysql.cj.jdbc.Driver";
+    public static Map<String, String> loadDatabase(StarRocksConfig config, List<String> dbNames) {
```
Comment (banmoy) [inline src/main/java/com/starrocks/connector/spark/sql/connect/StarRocksConnector.java:93]: The `dbNames` is confused. I guess it's the `namespaces` of spark. Actually you can `String.join(",", dbNames)` outside, and just deliver the final db name.

File: src/main/java/com/starrocks/connector/spark/sql/connect/StarRocksConnector.java [hunk 5]
```diff
@@ -71,21 +77,47 @@ public static StarRocksSchema getSchema(StarRocksConfig config) {
...
-    private static final String MAVEN_CENTRAL_URL = "https://repo1.maven.org/maven2/mysql/mysql-connector-java/";
+        for (Map<String, String> db : dbs) {
+            String dbName = Optional.ofNullable(db.get("SCHEMA_NAME"))
+                    .orElseThrow(() -> new StarrocksException("get Database SCHEMA_NAME error"));
+            return new DatabaseSpec(dbName).toJavaMap();
+        }
+        throw new StarrocksException("has no this database: " + config.getDatabase());
+    }
+
+    public static Map<String, String> getTables(StarRocksConfig config, List<String> dbNames) {
```
Comment (banmoy) [inline src/main/java/com/starrocks/connector/spark/sql/connect/StarRocksConnector.java:105]: `dbNames` is also confused, and should make it clear

Comment (banmoy) [inline src/main/java/com/starrocks/connector/spark/sql/connect/StarRocksConnector.java:105]: How about use List<org.apache.spark.sql.connector.catalog.Identifier> as the return?

File: src/main/java/com/starrocks/connector/spark/sql/StarRocksDataSourceProvider.java [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+
+        return new StarrocksRelation(sqlContext, transParameters);
+    }
+
+    @Override
+    public String shortName() {
+        return "starrocks";
+    }
+
+    //TODO 将所有转换的逻辑统一
```
Comment (banmoy) [inline src/main/java/com/starrocks/connector/spark/sql/StarRocksDataSourceProvider.java:75]: use english

File: src/main/java/com/starrocks/connector/spark/sql/StarRocksDataSourceProvider.java [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+            if (null == tbIdentifier || tbIdentifier.namespace().length == 0) {
+                LOG.error("request params must contain table identifier");
+                return null;
+            } else {
+                properties.setProperty(STARROCKS_TABLE_IDENTIFIER, identifierFullName(tbIdentifier));
+            }
+        }
+        Map<String, String> options = makeWriteCompatibleWithRead(config.getOriginOptions());
+        options.forEach((k, v) -> properties.setProperty(k, v));
+        Schema schema = SchemaUtils.discoverSchemaFromFe(properties);
```
Comment (banmoy) [inline src/main/java/com/starrocks/connector/spark/sql/StarRocksDataSourceProvider.java:109]: why not use the schema from jdbc?

File: src/main/java/com/starrocks/connector/spark/rest/RestService.java [hunk 8]
```diff
@@ -242,6 +243,7 @@ public static Schema getSchema(Settings cfg, Logger logger)
     public static Schema parseSchema(String response, Logger logger) throws StarrocksException {
         logger.trace("Parse response '{}' to schema.", response);
         ObjectMapper mapper = new ObjectMapper();
+        mapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
```
[the remark this answers is not in the document]
Comment (Jay-ju) [reply src/main/java/com/starrocks/connector/spark/rest/RestService.java:246]: here schema only owned column type,  but partition reader also need  EtlJobConfig.EtlTable  to get Partitions,  also need tableId .
so here, i want to struct include all those elments through one rest api like this.

public class Schema {
    private int status = 0;
    private List<Field> properties;
    private EtlJobConfig.EtlTable etlTable;
    private Long tableId;
}

File: src/main/java/com/starrocks/connector/spark/sql/conf/WriteStarRocksConfig.java [hunk 9]
```diff
@@ -46,6 +47,9 @@ public class WriteStarRocksConfig extends StarRocksConfigBase {
 
     private static final long serialVersionUID = 1L;
 
+    @VisibleForTesting
+    public static final boolean FOR_TEST_LOAD_WRITER = false;
```
[the remark this answers is not in the document]
Comment (Jay-ju) [reply src/main/java/com/starrocks/connector/spark/sql/conf/WriteStarRocksConfig.java:51]: delete

File: src/test/java/com/starrocks/connector/spark/sql/CatalogTest.java (hunk 1 above)
Comment (Jay-ju) [reply src/test/java/com/starrocks/connector/spark/sql/CatalogTest.java:1]: fixed

**Pull request closed without merging**